### PR TITLE
ENH: added analytical formulas for entropies

### PR DESCRIPTION
**Pull request by MatteoRaso**
#### Additional information

Adds analytical formulas for entropies (#17748)

### CODE REVIEW

**Review by tupui — CHANGES_REQUESTED** [review]
Thank you @MatteoRaso! Can you add documentation and tests? You can have a look at previous PRs linked in the issue to see what is needed. It's not much really.

@dschmitz89 would you like to review?

**Review by dschmitz89 — COMMENTED** [review]
Hi @MatteoRaso , thanks for finding those formulas and picking this up!

Sorry for the double work, but `beta`, `chi` and `chi2` were already implemented in #17742. They can be removed here then.

About the implementation itself: many of these formulas first use log(exp(something)). I would suggest (see one of the suggestions) to simplify these expressions symbolically and then perform the calculation directly in logspace.

About tests: for a first test that the implementation is probably right, i would compare with the current results from the main branch using the numerical integration (see for example the plots in #17742 ). Then, for unit testing some comparison with a multiprecision library is needed. In #17758 , an example was given to compute the entropy integral using [mpmath](https://github.com/scipy/scipy/pull/17758#discussion_r1068346336).

One last tip regarding the lint errors: they can be found locally by running `python dev.py lint`.

**Comment by MatteoRaso** [comment]
>About the implementation itself: many of these formulas first use log(exp(something)). I would suggest (see one of the suggestions) to simplify these expressions symbolically and then perform the calculation directly in logspace.

Good idea, done.

>About tests: for a first test that the implementation is probably right, i would compare with the current results from the main branch using the numerical integration (see for example the plots in https://github.com/scipy/scipy/pull/17742 ).

Is this necessary? ```test_continuous_basic``` checks entropies, so between that and the high-precision testing, I think there's enough test coverage.

>Then, for unit testing some comparison with a multiprecision library is needed. In https://github.com/scipy/scipy/pull/17758 , an example was given to compute the entropy integral using https://github.com/scipy/scipy/pull/17758#discussion_r1068346336.

Done. This is only tangential to the PR, but I noticed that the dgamma distribution doesn't have any tests. I'm beginning to add tests for a separate PR, but I just want to put that out there in case anybody was confused why the dgamma test suite only has the entropy test.

>One last tip regarding the lint errors: they can be found locally by running python dev.py lint.

I ran that, and it said just about all of the lint failures were from code other than what I wrote. The scipy development workflow guide says to only focus on your own linting problems, but that wouldn't solve the other linting problems that are causing the lint check to fail.

File: scipy/stats/tests/test_distributions.py [hunk 1]
```diff
@@ -3419,6 +3438,23 @@ def test_isf(self):
...
+    #   h = mp.log(A)
+    #   return h
+
+    @pytest.mark.parametrize("a, ref",
+    [(1.5, 2.0541199559354118267922018421759071394202272087878735),
+    (1.3, 1.9357296377121247644586320317632592032856887302320247),
+    (1.1, 1.7856502333412133408733521579693880954302277243835128)])
+
+
+    def test_entropy(self, a, ref):
```
Comment (dschmitz89) [inline scipy/stats/tests/test_distributions.py:3456]:
```suggestion
    @pytest.mark.parametrize("a, ref",
    [(1.5, 2.0541199559354118267922018421759071394202272087878735),
    (1.3, 1.9357296377121247644586320317632592032856887302320247),
    (1.1, 1.7856502333412133408733521579693880954302277243835128)])
    def test_entropy(self, a, ref):
```

File: scipy/stats/tests/test_distributions.py [hunk 2]
```diff
@@ -7284,6 +7320,25 @@ def test_sf_isf(self):
...
+
+    @pytest.mark.parametrize("m, ref",
+    [(5, -0.097341814372151976585790165774491299055778891485149),
+    (0.5, 0.72579135264472743236309761494744107178589733927753),
+    (10, -0.43426184310934903511376791552032937088827239690443)])
+
+
+    def test_entropy(self, m, ref):
+        assert_allclose(stats.nakagami.entropy(m), ref)
+
```
Comment (dschmitz89) [inline scipy/stats/tests/test_distributions.py:7341]:
```suggestion
    @pytest.mark.parametrize("m, ref",
    [(5, -0.097341814372151976585790165774491299055778891485149),
    (0.5, 0.72579135264472743236309761494744107178589733927753),
    (10, -0.43426184310934903511376791552032937088827239690443)])
    def test_entropy(self, m, ref):
        assert_allclose(stats.nakagami.entropy(m), ref)

```

File: scipy/stats/tests/test_distributions.py [hunk 3]
```diff
@@ -1108,6 +1108,25 @@ def test_9902_moments(self):
...
+    #   B = (a * mp.digamma(a) - b * mp.digamma(b)) / (2 * Z)
+    #   h = A + B
+    #   return h
+
+    @pytest.mark.parametrize("a, b, ref",
+    [(3, 4, -3.0638502581111376587051473653933044182228540802750831),
+    (3.6, 3.7, -5.6922960954665466660510221678517294674305674032403609),
+    (1, 3, 0.70921408458052569749601757859156839712816078772689478)])
+
+    def test_entropy(self, a, b, ref):
```
Comment (dschmitz89) [inline scipy/stats/tests/test_distributions.py:1127]:
```suggestion
    @pytest.mark.parametrize("a, b, ref",
    [(3, 4, -3.0638502581111376587051473653933044182228540802750831),
    (3.6, 3.7, -5.6922960954665466660510221678517294674305674032403609),
    (1, 3, 0.70921408458052569749601757859156839712816078772689478)])
    def test_entropy(self, a, b, ref):
```

File: scipy/stats/_continuous_distns.py [hunk 4]
```diff
@@ -1520,6 +1520,10 @@ def _ppf(self, q, a):
         fac = sc.gammainccinv(a, 1-abs(2*q-1))
         return np.where(q > 0.5, fac, -fac)
 
+    def _entropy(self, a):
+        h = np.log(2) + sc.loggamma(a) + a + (1 - a) * sc.digamma(a)
```
Comment (dschmitz89) [inline scipy/stats/_continuous_distns.py:1524]:
```suggestion
        h = np.log(2) + sc.gammaln(a) + a + (1 - a) * sc.digamma(a)
```

`loggamma` is in my understanding mostly useful for complex numbers even if the result is the same here.

File: scipy/stats/_continuous_distns.py [hunk 5]
```diff
@@ -6403,6 +6407,13 @@ def _sf(self, x, nu):
     def _isf(self, p, nu):
         return np.sqrt(1/nu * sc.gammainccinv(nu, p))
 
+    def _entropy(self, nu):
+        A = sc.loggamma(nu) - np.log(2)
```
Comment (dschmitz89) [inline scipy/stats/_continuous_distns.py:6411]:
```suggestion
        A = sc.gammaln(nu) - np.log(2)
```

Same as above.

**Review by dschmitz89 — CHANGES_REQUESTED** [review]

**Comment by dschmitz89** [comment]
> > About tests: for a first test that the implementation is probably right, i would compare with the current results from the main branch using the numerical integration (see for example the plots in #17742 ).
> 
> Is this necessary? `test_continuous_basic` checks entropies, so between that and the high-precision testing, I think there's enough test coverage.

The rationales behind high-precision testing based on numerical integration from the [discussion](https://github.com/scipy/scipy/pull/17758#discussion_r1068346336) were:

* sometimes the analytical integral evaluations are questionable. In the case of the F distribution, the Wikipedia formula was plain wrong. If we just use the mpmath versions of the scipy implementation, we do not test this.
* `test_continuous_basic` does not actually test the computed entropy values against a reference from what I see.

I think the code [snippet](https://github.com/scipy/scipy/pull/17758#discussion_r1068346336) for high precision integration could serve as a reference here too, just the pdf needs to be replaced.

File: scipy/stats/_continuous_distns.py [hunk 6]
```diff
@@ -8745,6 +8756,15 @@ def isf_right(q, a, b):
...
         return out
 
+    def _entropy(self, a, b):
+        A = sc.digamma(a)
+        B = sc.digamma(b)
+        Z = B - A
+        C = np.log(np.sqrt(2 * np.pi * np.e) * Z)
+        D = (a * A - b * B) / (2 * Z)
+        h = C + D
+        return h
```
Comment (dschmitz89) [inline scipy/stats/_continuous_distns.py:8766]:
```suggestion
        A = sc.ndtr(a)
        B = sc.ndtr(b)
        Z = B - A
        C = np.log(np.sqrt(2 * np.pi * np.e) * Z)
        D = (a * _norm_pdf(a) - b * _norm_pdf(b)) / (2 * Z)
        h = C + D
        return h
```

File: scipy/stats/_continuous_distns.py (hunk 4 above)
Comment (MatteoRaso) [reply scipy/stats/_continuous_distns.py:1524]: I don't think it really matters. From what I understand, using loggamma doesn't affect the speed or accuracy, so we might as well leave it.

File: scipy/stats/_continuous_distns.py (hunk 6 above)
Comment (MatteoRaso) [reply scipy/stats/_continuous_distns.py:8766]: This code doesn't work.

Comment (MatteoRaso) [reply scipy/stats/_continuous_distns.py:8766]: This code doesn't work.

**Comment by MatteoRaso** [comment]
>could you maybe post screenshots of the formulas from the book you used also for comparison? Currently, the math can actually only be reviewed for truncnorm :)

Yes, but with the caveat that the definitions have errors, albeit ones that can be easily corrected. I didn't mention it, but the book shows how the entropy formulas are derived, which might make the need for numerical integration redundant.
[https://imgur.com/a/t4S4y3z](url)

File: scipy/stats/_continuous_distns.py (hunk 6 above)
Comment (MatteoRaso) [reply scipy/stats/_continuous_distns.py:8766]: Thanks, I didn't realize I misread the definition.

**Comment by MatteoRaso** [comment]
>which might make the need for numerical integration redundant

Scratch that. I've looked over the math for the entropy derivations a bit more closely, and I think the author messed it up. I'm going to add the numerical integration, as well as review the derivations, but that'll take me a bit of time (probably no longer than a week).

**Comment by dschmitz89** [comment]
@MatteoRaso : would you consider opening separate PRs for the individual distributions when you have time to get back to this? 

I don't intend to scratch your work here, but that would make review a lot easier. Some of these expressions can become quite involved, and as you also noticed, references are not always reliable. In case one of the distributions you are working on is difficult, it would hold up the rest. Not talking about the general review backlog ;).

**Comment by MatteoRaso** [comment]
>@MatteoRaso : would you consider opening separate PRs for the individual distributions when you have time to get back to this?

Good idea. As you mentioned, I have been having problems with a few of the distributions, so I'll open PRs for the ones that I have done.

**Pull request closed without merging**